include/deal.II/matrix_free/fe_evaluation.h

```diff
@@ -2766,6 +2742,25 @@ namespace internal
     }
 
     template <typename VectorType>
+    void process_dof_gather (const unsigned int      *indices,
+                             VectorType              &vec,
+                             VectorizedArray<Number> &res,
+                             internal::bool2type<true>) const
+    {
+      res.gather(vec.begin(), indices);
+    }
+
+    template <typename VectorType>
+    void process_dof_gather (const unsigned int      *indices,
+                             VectorType              &vec,
+                             VectorizedArray<Number> &res,
+                             internal::bool2type<false>) const
```

Comment:
The problem is that we allow `FEEvaluation<...,double>` to read from `Vector<float>` and the other way around. However, there are no gather instructions going to different types, and our implementation of `VectorizedArray` does not support conversion between float and double. Since the situation is rare, I think the scalar variant is fine. To direct the compiler to the correct function, I use the `bool2type` variant.